NIST elliptic curves via mirage-crypto-ec

hannesm opened the pull request:
This PR integrated elliptic curves into this repository. It is based on [fiat](https://github.com/mirage/fiat), and due to some dependencies for ECDSA (compute `k` in a deterministic way, hash algorithms) it has been imported into this repository.

The API (Mirage_crypto_ec) is not yet set into stone (multiple NIST curves should be reflected, documentation is lacking). It is work in progress, with the main goal here to test with the various CI systems in place.

The extraction process (from [fiat-crypto](https://github.com/mit-plv/fiat-crypto)) is done via `GNUmakefile` in `ec/native`. Note that it has been updated, and the fiat-crypto provided inversion is used now (previously, there was a `fe_inv`). The test cases pass, though ECDSA may need some `Cstruct.rev` adaptions, and more tests (from wycheproof) are needed for ECDSA.

Hope that @NathanReb @emillon @pascutto are fine with the code import here.

Sponsored by: Nitrokey GmbH

JasonGross commented:
> The extraction process (from [fiat-crypto](https://github.com/mit-plv/fiat-crypto)) is done via `GNUmakefile` in `ec/native`.

If you're already using dune and OCaml, perhaps you want to depend on [our opam package](https://github.com/coq/opam-coq-archive/blob/master/extra-dev/packages/coq-fiat-crypto/coq-fiat-crypto.dev/opam) which installs the relevant binaries to `%{bin}%`?  I'm happy to release a non-dev version of fiat-crypto if you'd prefer that, or update the opam package if it doesn't currently meet your needs.

hannesm commented:
@JasonGross thanks for your pointer to the opam package. This sounds like a nice way forward on documenting (and running) how to re-do the extraction. For the common OCaml user, I don't think it is reasonable to generate the C code themselves (but instead we'll ship pre-generated C code).

JasonGross commented:
> For the common OCaml user, I don't think it is reasonable to generate the C code themselves (but instead we'll ship pre-generated C code

Indeed, I was just suggesting that perhaps your makefile setup can be configured for the opam-based setup rather than assuming fiat-crypto in a parent directory.

dinosaure commented on mirage-crypto-ec.opam:
```diff
@@ -0,0 +1,56 @@
+  "ocaml" {>= "4.08.0"}
+  "cstruct" {>= "3.5.0"}
+  "dune-configurator"
+  "eqaf" {>= "0.5"}
```
It's better to depend on `0.7`.

hannesm replied:
thanks, done

avsm commented on .cirrus.yml:
```diff
@@ -1,14 +1,23 @@
+task:
+  << : *PKG_INSTALL_TASK_TEMPLATE
+  ocaml_script: opam init -a --comp=4.10.2
+  dependencies_script: eval `opam env` && opam pin add -y mirage-crypto.dev . && opam pin add -y mirage-crypto-rng.dev . && opam pin add -y mirage-crypto-pk.dev . && opam pin add -y mirage-crypto-ec.dev . && opam install -y -t --deps-only mirage-crypto mirage-crypto-rng mirage-crypto-pk mirage-crypto-ec
+  build_script: eval `opam env` && opam exec -- dune build -p mirage-crypto,mirage-crypto-rng,mirage-crypto-pk,mirage-crypto-ec
+  test_script: eval `opam env` && opam exec -- dune runtest -p mirage-crypto,mirage-crypto-rng,mirage-crypto-pk,mirage-crypto-ec
```
minor: the eval is not necessary here as the various opam commands will add the right env variables

hannesm commented:
current status:
- [x] P-224 (ECDH & ECDSA & test cases from RFC 6979 and wycheproof, 32 and 64 bit)
- [x] P-256 (ECDH & ECDSA & test cases from RFC 6979 and wycheproof, 32 and 64 bit)
- [x] P-384 (ECDH & ECDSA & test cases from RFC 6979 and wycheproof, 32 and 64 bit)
- [x] P-521 (ECDH & ECDSA & test cases from wycheproof, 64 bit only, RFC 6979 test cases fail since `k` is computed differently (bit_length % 8 <> 0)

Left items:
- [ ] SECP256K1 (needs more thorough investigation, likely scalar multiplication is wrong)
- [ ] validate API (by using it in X509 & TLS & ...)
- [ ] write documentation for public API

Delayed until later:
- [ ] P521 32 bit code (will be delayed until someone bothers to work on it)
- [ ] RFC6979 test cases for P521 (will be delayed until someone has a good idea avoiding bit-fiddling)

Reviews / comments / feedback appreciated :) [a different step / PR is to integrate 25519 (ECDH & EdDSA) and 448 (ECDH & EdDSA)]

hannesm commented on ec/native/p224_stubs.c:
```diff
@@ -0,0 +1,143 @@
+
+#include <caml/memory.h>
+
+CAMLprim value mc_p224_sub(value out, value a, value b)
+{
+	CAMLparam3(out, a, b);
+	fiat_p224_sub(Caml_ba_data_val(out), Caml_ba_data_val(a), Caml_ba_data_val(b));
+	CAMLreturn(Val_unit);
```
Uhm, this is done at the OCaml declaration-side (e.g. for P-224 at https://github.com/mirage/mirage-crypto/pull/101/files#diff-b65c54aec12f050b8540b66b5b49e4fc2289e3137ec35bd9a7ef23f9dec38ec8R611-R638). Anything else that is needed there, or any occassion I forgot to mark `[@@noalloc]`?

avsm replied:
Essentially the question above: if these are computationally intensive and could take some time, then we should drop the runtime lock and not make them noalloc.

dinosaure replied:
> Essentially the question above: if these are computationally intensive and could take some time, then we should drop the runtime lock and not make them noalloc.

It should be easy due to the use of non-relocatable bigarrays.

JasonGross replied:
CPU-intensive is a comparative term.  Certainly these functions are the CPU-hotspots in performing curve operations, but the time spent in any single invocation is quite small.  The are some numbers in [Andres Erbsen, Jade Philipoom, Jason Gross, Robert Sloan, Adam Chlipala. Simple High-Level Code For Cryptographic Arithmetic -- With Proofs, Without Compromises. Proceedings of the IEEE Symposium on Security & Privacy 2019 (S&P'19). May 2019.](http://adam.chlipala.net/papers/FiatCryptoSP19/FiatCryptoSP19.pdf) The table on page 10 shows that running the entire scalar multiplication (which will call these functions a handful of times) takes about 1k--3k CPU cycles.  The plot in in Fig. 10 corresponds to 1000 sequential
computations of a 256-bit Montgomery ladder, and even the slowest (~500 bit primes on 32-bit machines) amount to less than 0.001--0.01 seconds per call.

Additionally, these primitives are low-level, intended to be used to implement higher-level curve operations.  I expect that the boundary of these functions is not the place to be releasing runtime locks, and that should instead happen at the boundary of the curve-level operations (which, I believe, should also not be allocating anything on the heap).  Another point in this direction is that ideally we'd like to generate the curve-level operations in C as well and expose that API, rather than this lower-level one.

hannesm commented:
@EduardoRFS any idea what the root cause of the cross compile failures is? are they temporary?

hannesm commented:
this is ready to be merged: some initial documentation is included in this PR, also the P521 32bit code has been generated and added. I squashed the development commits together, and am fine with the remaning ones.

from the comment above, future PRs may:
- add more curves (SECP_K1, require a different implementation of scalar_mult)
- reduce allocations (or in general, improve performance)
- adapt 'deterministic k' to use the bit size of the curve (and have p521 rfc6979 tests not fail)
- add edwards curves (25519, 448) with dh and eddsa

EduardoRFS commented:
@hannesm yeah they're, github actions connections seems to be failing, will be addressing it tomorrow, feel free to ignore it
